README
WWW::FreshMeat::API version 0.01

This distribution includes the following modules:

    WWW::FreshMeat::API                   (0.01)
    WWW::FreshMeat::API::Session          (0.01)
    WWW::FreshMeat::API::Pub              (0.01)
    WWW::FreshMeat::API::Pub::V1_03       (0.01)
    WWW::FreshMeat::API::Agent::XML::RPC  (0.01)


DESCRIPTION

FreshMeat (http://freshmeat.net) provides a very simple XML-RPC API which allows a user to inspect what projects
the user as uploaded and also provides an update & withdrawal mechanism of the users projects.

Requirements....

    1) Must have a FreshMeat login & password

    2) Must have already loaded project onto http://freshmeat.net

For now see ./examples/freshmeat-submit.pl.

More details to follow.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


DEPENDENCIES

This module requires these other modules and libraries:

	Moose
    XML::RPC
    

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WWW::FreshMeat::API


COPYRIGHT AND LICENCE

Copyright (C) 2009 Barry Walsh (Draegtun Systems Ltd)

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for WWW-FreshMeat-API

0.01    Sat January 17, 2009
    * First version, released on an unsuspecting world.
      - alpha version... but it does work!
      - turned into CPAN module from dirty prototypes (see initial git commits)
      
0.02    Sun July 19, 2009
    * Service announcement... Freshmeat have dropped the XML-RPC API
      - Made annoucement changes to Readme & API.pm

prototype/FMdata.pm

package FMdata;
use base 'Exporter';
our @EXPORT_OK = qw/get_api_info/;

our $api = {

    fetch_available_licenses => {
        desc    => 'Fetch all available licenses',
        params  => [],
        returns => [ qw/Array of available licenses/ ],
    },
        
    fetch_available_release_foci => {
        desc    => 'Fetch all available release focus types',
        params  => [],
        returns => [ qw/Struct of available release focus types and associated IDs/ ],
    },
    
    fetch_branch_list => {
        desc    => 'Fetch all branch names and IDs for a given project', 
        params  => [ qw/SID project_name/ ],
        returns => [ qw/Array of branch name strings./ ],
    },
        
    fetch_project_list => {
        desc    => 'Fetch all projects assigned to logged in user',
        params  => [ qw/SID/ ],
        returns => [ qq{Array of structs consisting of "projectname_full", "projectname_short", "project_status", and "project_version"} ],
    },
        
    fetch_release => {
        desc    => 'Fetch data from a pending release submission',
        params  => [ qw/SID project_name branch_name version/ ],
        returns => [ qq{Struct consisting of "version", "changes", "release_focus", and "hide_from_frontpage"} ],
    },
    
    login => {
        desc    => 'Start an XML-RPC session',
        params  => [ qw/username password/ ],
        returns => [ 'SID', 'Lifetime', 'API Version' ],
    },
    
    logout => {
        desc    => 'End an XML-RPC session',
        params  => [ qw/SID/ ],
        returns => [ qq{Struct of "OK" => "Logout successful." if logout was successful} ],
    },
    
    publish_release	=> {
        desc    => 'Publish a new release',
        params  => [ qw/
            SID project_name branch_name version changes release_focus hide_from_frontpage
            license url_homepage url_tgz url_bz2 url_zip url_changelog url_rpm url_deb 
            url_osx url_bsdport url_purchase url_cvs url_list url_mirror url_demo	   
        / ],
        returns => [ qq{Struct of "OK" => "submission successful"} ],
    },
    
    withdraw_release => {
        desc    => 'Take back a release submission',
        params  => [ qw/SID project_name branch_name version/ ],
        returns => [ qq{Struct of "OK" => "Withdraw successful."} ],
    },
};

# [ Appendix A: Release focus IDs ]
# 0 - N/A
# 1 - Initial freshmeat announcement
# 2 - Documentation
# 3 - Code cleanup
# 4 - Minor feature enhancements
# 5 - Major feature enhancements
# 6 - Minor bugfixes
# 7 - Major bugfixes
# 8 - Minor security fixes
# 9 - Major security fixes
# 
# 
# [ Appendix B: Error codes ]
#  10 - Login incorrect
#  20 - Session inconsistency
#  21 - Session invalid
#  30 - Branch ID incorrect
#  40 - Permission to publish release denied
#  50 - Version string missing
#  51 - Duplicate version string
#  60 - Changes field empty
#  61 - Changes field too long
#  62 - Changes field contains HTML
#  70 - No valid email address set
#  80 - Release not found
#  81 - Project not found
#  90 - Release focus missing
#  91 - Release focus invalid
# 100 - License invalid
# 999 - Unknown error

sub get_api_info { $api }

1;
    


prototype/freshmeat.pl
#!/usr/local/bin/perl

# second prototype - 12th January 2009

use strict;
use warnings;
use Perl6::Say;

{
    package FreshMeat::Base;
    use Moose::Role;
    
    has 'session'   => ( isa => 'HashRef', is => 'rw', default => sub { +{} } );
    has 'fatal' => ( isa => 'Bool', is => 'rw', default => 1 );
    
    sub dump_session { $_[0]->{session} }
    
    sub sid { $_[0]->session->{ SID } }
    
    sub lifetime { $_[0]->session->{ Lifetime } }
    
    sub api { $_[0]->session->{ 'API Version' } }
    
    no Moose::Role;
}

{
    package FreshMeat::API::XML::RPC;
    use Moose::Role;
    use XML::RPC;
    
    requires 'session';
    
    has 'agent' => ( 
        isa      => 'XML::RPC', 
        is       => 'ro', 
        default  => sub { XML::RPC->new( 'http://freshmeat.net/xmlrpc/' ) },
        handles  => [ qw/call/ ],
        lazy     => 1,
        required => 1,
    );
    
    sub login {
        my ( $self, %params ) = @_;
        $self->session( $self->agent->call( 'login', \%params ) );
    }
        
    sub fetch_project_list {
        my $self = shift;
        $self->agent->call( 'fetch_project_list' );
    }
    
    no Moose::Role;
}

{
    package FreshMeat::API::MockObject;
    use Moose;
    use Data::Dumper;
    
    sub call {
        print "Mockery $_[1]\n";
        return "Mockery $_[1] " . Dumper( $_[2] );
    }
}

{
    package FreshMeatAPIV1_03;
    use base 'Exporter';
    our @EXPORT_OK = qw/get_api_details/;
    
    our $api = {
    
        fetch_available_licenses => {
            desc    => 'Fetch all available licenses',
            params  => [],
            returns => [ qw/Array of available licenses/ ],
        },
            
        fetch_available_release_foci => {
            desc    => 'Fetch all available release focus types',
            params  => [],
            returns => [ qw/Struct of available release focus types and associated IDs/ ],
        },
        
        fetch_branch_list => {
            desc    => 'Fetch all branch names and IDs for a given project', 
            params  => [ qw/SID project_name/ ],
            returns => [ qw/Array of branch name strings./ ],
        },
            
        fetch_project_list => {
            desc    => 'Fetch all projects assigned to logged in user',
            params  => [ qw/SID/ ],
            returns => [ qq{Array of structs consisting of "projectname_full", "projectname_short", "project_status", and "project_version"} ],
        },
            
        fetch_release => {
            desc    => 'Fetch data from a pending release submission',
            params  => [ qw/SID project_name branch_name version/ ],
            returns => [ qq{Struct consisting of "version", "changes", "release_focus", and "hide_from_frontpage"} ],
        },
        
        login => {
            desc    => 'Start an XML-RPC session',
            params  => [ qw/username password/ ],
            returns => [ 'SID', 'Lifetime', 'API Version' ],
        },
        
        logout => {
            desc    => 'End an XML-RPC session',
            params  => [ qw/SID/ ],
            returns => [ qq{Struct of "OK" => "Logout successful." if logout was successful} ],
        },
        
        publish_release	=> {
            desc    => 'Publish a new release',
            params  => [ qw/
                SID project_name branch_name version changes release_focus hide_from_frontpage
                license url_homepage url_tgz url_bz2 url_zip url_changelog url_rpm url_deb 
                url_osx url_bsdport url_purchase url_cvs url_list url_mirror url_demo	   
            / ],
            returns => [ qq{Struct of "OK" => "submission successful"} ],
        },
        
        withdraw_release => {
            desc    => 'Take back a release submission',
            params  => [ qw/SID project_name branch_name version/ ],
            returns => [ qq{Struct of "OK" => "Withdraw successful."} ],
        },
    };
    
    # [ Appendix A: Release focus IDs ]
    # 0 - N/A
    # 1 - Initial freshmeat announcement
    # 2 - Documentation
    # 3 - Code cleanup
    # 4 - Minor feature enhancements
    # 5 - Major feature enhancements
    # 6 - Minor bugfixes
    # 7 - Major bugfixes
    # 8 - Minor security fixes
    # 9 - Major security fixes
    # 
    # 
    # [ Appendix B: Error codes ]
    #  10 - Login incorrect
    #  20 - Session inconsistency
    #  21 - Session invalid
    #  30 - Branch ID incorrect
    #  40 - Permission to publish release denied
    #  50 - Version string missing
    #  51 - Duplicate version string
    #  60 - Changes field empty
    #  61 - Changes field too long
    #  62 - Changes field contains HTML
    #  70 - No valid email address set
    #  80 - Release not found
    #  81 - Project not found
    #  90 - Release focus missing
    #  91 - Release focus invalid
    # 100 - License invalid
    # 999 - Unknown error
    
    sub get_api_info { $api }
    
    1;
    
}

{
    package FreshMeat::API;
    use Moose::Role;
    requires 'session';
    
    use FMdata qw/get_api_info/;

    my %method = (
        login => 'login',
        fetch_project_list => 'fetch_project_list',
    );
    
    sub BUILD {
        my $self = shift;
        for my $name ( keys %{ get_api_info() } ) {
            $self->meta->add_method( $name => sub {
                my ( $self, %params ) = @_;
                $self->session( $self->agent->call( $name, \%params ) );
            });
        }
    }
    
    # sub BUILD {
    #     my $self = shift;
    #     for my $name ( keys %method ) {
    #         $self->meta->add_method( $name => sub {
    #             my ( $self, %params ) = @_;
    #             $self->session( $self->agent->call( $method{ $name }, \%params ) );
    #         });
    #     }
    # }
    
    no Moose::Role;
}

{
    package FreshMeat::API::Mock;
    use Moose::Role;
    requires 'session';
    with 'FreshMeat::API';
    
    has 'agent' => ( 
        isa      => 'Object', 
        is       => 'ro', 
        default  => sub { FreshMeat::API::MockObject->new },
        handles  => [ qw/call/ ],
        lazy     => 1,
        required => 1,
    );
    
    sub loginx {
        my ( $self, %params ) = @_;
        $self->session( $self->agent->call( 'login', \%params ) );
    }
    
}

{
    package FreshMeat;
    use Moose;
    with 'FreshMeat::Base', 'FreshMeat::API::Mock';
    

    no Moose;
}

use Data::Dumper;
my $fm = FreshMeat->new;
$fm->login( username => 'username', password => 'pass1' );
say Dumper( $fm );

$fm->publish_release;
#say $fm->sid;
#say Dumper( $fm->fetch_project_list );


__END__

use Data::Dumper;

{
    package FreshMeat::Agent::XML::RPC;
    use Moose;
    use XML::RPC;

    has '_url' => ( isa => 'Str', is => 'ro', default => 'http://freshmeat.net/xmlrpc/' );
    
    has '_agent' => ( 
        isa      => 'XML::RPC', 
        is       => 'ro', 
        default  => sub { XML::RPC->new( $_[0]->{_url} ) },
        handles  => [ qw/call/ ],
        lazy     => 1,
        required => 1,
    );

    no Moose;
}

{
    package FreshMeat;
    use Moose;
    use Data::Dumper;
    
    has 'session'   => ( isa => 'Str', is => 'rw', default => '' );
    has 'fatal' => ( isa => 'Bool', is => 'rw', default => 1 );
    has 'agent' => ( 
        isa     => 'Object',  
        is      => 'ro', 
        default => sub { FreshMeat::Agent::XML::RPC->new },
    );
    
    sub login {
        my ( $self, %params ) = @_;
        $self->session( $self->agent->call( 'login', \%params ) );
        print Dumper( $self->{session} );
    }
    
    sub sid { $_[0]->session->{SID} }
    
    no Moose;
}

my $fm = FreshMeat->new;
$fm->login( username => 'username', password => 'pass1' );
say $fm->sid;




__END__


my $fm = FreshMeat->new( fatal => 1, agent => FreshMeat::Agent::XML::RPC->new( 'http://freshmeat.net/xmlrpc/', api_version => 1 ) );

$fm->login( username => 'username', password => 'pass1' );

$fm->login( profile => '.freshmeat' );

$fm->login( use_netrc => 1 );

say $fm->sid;   # SID
say $fm->lifetime;   # in secs  Lifetime
say $fm->epoch;
say $fm->expire;
say $fm->time_left;  # in secs
say $fm->api;  # API Version

$fm->fetch_project_list;   # have login automatically do this?
$fm->get_projects;         # so can do this.. returns array of hashes



__END__

# initial prototype - 8th January 2008

my $fm = XML::RPC->new( 'http://freshmeat.net/xmlrpc/' );

my $session = $fm->call( 'login', { username => 'username', password => 'pass1x2' });

say Dumper( $session );

my $x = $fm->call( 'fetch_project_list', { SID => $session->{SID} } );

say $x;
say Dumper( $x );

__END__

HASH(0x196963c)
$VAR1 = {
          'SID' => 'cb351295f663432d194fc865f9393b4c',
          'Lifetime' => '600',
          'API Version' => '1.03'
        };
        
        
        
Login failure
$VAR1 = {
          'faultString' => 'Login incorrect',
          'faultCode' => '10'
        };


lib/WWW/FreshMeat/API/Pub.pm
package WWW::FreshMeat::API::Pub;
use Moose::Role;
use WWW::FreshMeat::API::Pub::V1_03 qw/get_api_info/;

our $VERSION = '0.01';

#requires 'sid';

sub BUILD {
    my $self = shift;
    my $api  = get_api_info();
    
    for my $name ( keys %{ $api } ) {
        
        # if mock then setup some dummy methods
        if ( $self->mock ) {
            $self->meta->add_method( $name => sub {
                my ( $self, %params ) = @_;
                return $api->{ $name }->{ returns };
            });
            next;
        }
         
        # else public methods... however next two require setting up session
        next  if $name eq 'login';
        next  if $name eq 'logout';
        
        # now all remaining public methods - v1.03
        if ( @{ $api->{ $name }->{ params } } ) {
            
            # build methods with params
            $self->meta->add_method( $name => sub {
                my ( $self, %params ) = @_;
                
                # add SID if it needs it (must be first in array)
                %params = ( SID => $self->session->{ SID }, %params )
                    if  $api->{ $name }->{ params }->[0] eq 'SID';
                    
                $self->agent->call( $name, \%params );
            });
        }
        
        else {
            # build method with no params
            $self->meta->add_method( $name => sub {
                my $self = shift;
                $self->agent->call( $name );
            });
        }
    }   
}

# session methods
sub login {
    my ( $self, %params ) = @_;
    $self->session( $self->agent->call( 'login', \%params ) );
}

sub logout {
    my $self = shift;
    $self->agent->call( 'logout', { SID => $self->session->{ SID } } );
    $self->clear_session;
}

no Moose::Role;

1;


__END__


=head1 NAME

WWW::FreshMeat::API::Pub - FreshMeats published / public (take your pick!) API methods

=head1 VERSION

Version 0.01


=head1 SYNOPSIS

    use Moose;
    
    with 'WWW::FreshMeat::API::Pub';


=head1 DESCRIPTION

This is a Moose role which builds the public/published FreshMeat API from WWW::FreshMeat::API::Pub::V1_03 metadata.
    

=head1 EXPORT

None.


=head1 METHODS

See FreshMeat API docs.  Methods use same names.  
Where docs say returns "Array" then a ArrayRef is return.
Where docs says returns "Struct" then HashRef is returned.
So if docs say "Array of structs" then ArrayRef of HashRefs is returned by said method.

=head2 fetch_available_licenses

=head2 fetch_available_release_foci

=head2 fetch_branch_list

=head2 fetch_project_list

=head2 fetch_release

=head2 publish_release

=head2 withdraw_release

=head2 login

=head2 logout


=head1 AUTHOR

Barry Walsh, C<< <draegtun at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-freshmeat-api at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FreshMeat-API>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FreshMeat::API::Pub


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FreshMeat-API>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FreshMeat-API>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FreshMeat-API>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FreshMeat-API/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 SEE ALSO

=head2 Freshmeat API FAQ

http://freshmeat.net/faq/view/49/

=head2 Freshmeat XML-RPC API announcement

http://freshmeat.net/articles/view/1048/

=head2 Other WWW::FreshMeat::API modules

L<WWW::FreshMeat::API>




=head1 COPYRIGHT & LICENSE

Copyright 2009 Barry Walsh (Draegtun Systems Ltd), all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



lib/WWW/FreshMeat/API/Session.pm
package WWW::FreshMeat::API::Session;
use Moose::Role;

our $VERSION = '0.01';

has 'session' => ( isa => 'HashRef', is => 'rw', default => sub { +{} } );
has 'fatal'   => ( isa => 'Bool', is => 'rw', default => 1 );

sub sid { $_[0]->session->{ SID } }

sub lifetime { $_[0]->session->{ Lifetime } }

sub api { $_[0]->session->{ 'API Version' } }

sub clear_session { 
    $_[0]->{session} = {};
}


no Moose::Role;

1;


__END__


=head1 NAME

WWW::FreshMeat::API::Session - Session role for WWW::FreshMeat::API

=head1 VERSION

Version 0.01


=head1 SYNOPSIS

    use Moose;

    with 'WWW::FreshMeat::API::Session';
    

=head1 DESCRIPTION

This is a Moose role used by WWW::FreshMeat::API providing session attributes & methods.


=head1 EXPORT

None.


=head ATTRIBUTES

=head2 session

=head2 fatal

Not implemented yet.


=head1 METHODS

=head2 sid

=head2 lifetime

=head2 api

=head2 clear_session


=head1 AUTHOR

Barry Walsh, C<< <draegtun at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-freshmeat-api at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FreshMeat-API>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FreshMeat::API::Session


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FreshMeat-API>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FreshMeat-API>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FreshMeat-API>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FreshMeat-API/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 SEE ALSO

=head2 Other WWW::FreshMeat::API modules

L<WWW::FreshMeat::API>


=head1 COPYRIGHT & LICENSE

Copyright 2009 Barry Walsh (Draegtun Systems Ltd), all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



lib/WWW/FreshMeat/API/Pub/V1_03.pm
package WWW::FreshMeat::API::Pub::V1_03;
use base 'Exporter';
our @EXPORT_OK = qw/get_api_info/;
our $VERSION = '0.01';

# pulled by testing API....
our $api = {

    fetch_available_licenses => {
        desc    => 'Fetch all available licenses',
        params  => [],
        returns => [
                  'Academic Free License (AFL)',
                  'Adaptive Public License (APL)',
                  'Affero General Public License',
                  'Aladdin Free Public License (AFPL)',
                  'Apple Public Source License (APSL)',
                  'Artistic License',
                  'Boost Software License',
                  'BSD License (original)',
                  'BSD License (revised)',
                  'Common Development and Distribution License (CDDL)',
                  'Common Public License',
                  'Copyback License',
                  'DFSG approved',
                  'Eclipse Public License',
                  'Educational Community License',
                  'Eiffel Forum License (EFL)',
                  'Free For Educational Use',
                  'Free For Home Use',
                  'Free for non-commercial use',
                  'Free To Use But Restricted',
                  'Freely Distributable',
                  'Freeware',
                  'GNAT Modified GPL (GMGPL)',
                  'GNU Free Documentation License (FDL)',
                  'GNU General Public License (GPL)',
                  'GNU General Public License v2',
                  'GNU General Public License v3',
                  'GNU Lesser General Public License (LGPL)',
                  'Guile license',
                  'IBM Public License',
                  'MIT/X Consortium License',
                  'MITRE Collaborative Virtual Workspace License (CVW)',
                  'Mozilla Public License (MPL)',
                  'Netscape Public License (NPL)',
                  'Nokia Open Source License (NOKOS)',
                  'Open Software License',
                  'OSI Approved',
                  'Other/Proprietary License',
                  'Other/Proprietary License with Free Trial',
                  'Other/Proprietary License with Source',
                  'Perl License',
                  'Public Domain',
                  'Python License',
                  'Q Public License (QPL)',
                  'Ricoh Source Code Public License',
                  'Shareware',
                  'SUN Binary Code License',
                  'SUN Community Source License',
                  'SUN Public License',
                  'The Apache License',
                  'The Apache License 2.0',
                  'The CeCILL License',
                  'The Clarified Artistic License',
                  'The Latex Project Public License (LPPL)',
                  'The Open Content License',
                  'The PHP License',
                  'Voxel Public License (VPL)',
                  'W3C License',
                  'WTFPL',
                  'zlib/libpng License',
                  'Zope Public License (ZPL)',
            ],
    },
        
    fetch_available_release_foci => {
        desc    => 'Fetch all available release focus types',
        params  => [],
        returns => {
                  'N/A' => '0',
                  'Major security fixes' => '9',
                  'Documentation' => '2',
                  'Minor security fixes' => '8',
                  'Initial freshmeat announcement' => '1',
                  'Minor bugfixes' => '6',
                  'Code cleanup' => '3',
                  'Major bugfixes' => '7',
                  'Minor feature enhancements' => '4',
                  'Major feature enhancements' => '5'
        },
    },
    
    fetch_branch_list => {
        desc    => 'Fetch all branch names and IDs for a given project', 
        params  => [ qw/SID project_name/ ],
        returns => [],  # returns list of branches, for eg. 'Default'
    },
        
    fetch_project_list => {
        desc    => 'Fetch all projects assigned to logged in user',
        params  => [ qw/SID/ ],
        returns => [ { projectname_full => 1, projectname_short => 1, project_status => 1, project_version => 1 } ],
    },
        
    fetch_release => {
        desc    => 'Fetch data from a pending release submission',
        params  => [ qw/SID project_name branch_name version/ ],
        returns => { version => 1, changes => 1, release_focus => 1, hide_from_frontpage => 1 },
    },
    
    login => {
        desc    => 'Start an XML-RPC session',
        params  => [ qw/username password/ ],
        returns => { 'SID' => 1, 'Lifetime' => 1, 'API Version' => 1 },
    },
    
    logout => {
        desc    => 'End an XML-RPC session',
        params  => [ qw/SID/ ],
        returns => { OK => "Logout successful." },  #  if logout was successful
    },
    
    publish_release	=> {
        desc    => 'Publish a new release',
        params  => [ qw/
            SID project_name branch_name version changes release_focus hide_from_frontpage
            license url_homepage url_tgz url_bz2 url_zip url_changelog url_rpm url_deb 
            url_osx url_bsdport url_purchase url_cvs url_list url_mirror url_demo	   
        / ],
        returns => { OK => "submission successful" },  # if successful!
    },
    
    withdraw_release => {
        desc    => 'Take back a release submission',
        params  => [ qw/SID project_name branch_name version/ ],
        returns => { OK => "Withdraw successful." },  # if successfully withdrawn!
    },
};


# Below is only bit of info I actually used from freshmeat-submit
# [ Appendix A: Release focus IDs ]
# 0 - N/A
# 1 - Initial freshmeat announcement
# 2 - Documentation
# 3 - Code cleanup
# 4 - Minor feature enhancements
# 5 - Major feature enhancements
# 6 - Minor bugfixes
# 7 - Major bugfixes
# 8 - Minor security fixes
# 9 - Major security fixes
# 
# 
# [ Appendix B: Error codes ]
#  10 - Login incorrect
#  20 - Session inconsistency
#  21 - Session invalid
#  30 - Branch ID incorrect
#  40 - Permission to publish release denied
#  50 - Version string missing
#  51 - Duplicate version string
#  60 - Changes field empty
#  61 - Changes field too long
#  62 - Changes field contains HTML
#  70 - No valid email address set
#  80 - Release not found
#  81 - Project not found
#  90 - Release focus missing
#  91 - Release focus invalid
# 100 - License invalid
# 999 - Unknown error

sub get_api_info { $api }

1;
    
__END__

=head1 NAME

package WWW::FreshMeat::API::Pub::V1_03 - Metadata for FreshMeat API v1.03

=head1 VERSION

Version 0.01


=head1 SYNOPSIS

    use package WWW::FreshMeat::API::Pub::V1_03 qw/get_api_info/;
    

=head1 DESCRIPTION

Exports on requestion get_api_info() which is simple a sub around hashref metadata.

Used by WWW::FreshMeat::API::Pub to build the published API as methods.

=head1 EXPORT

=head2 get_api_info()

On request


=head1 FUNCTIONS

=head2 get_api_info()


=head1 AUTHOR

Barry Walsh, C<< <draegtun at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-freshmeat-api at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FreshMeat-API>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.



=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FreshMeat::API::Pub::V1_03


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FreshMeat-API>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FreshMeat-API>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FreshMeat-API>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FreshMeat-API/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 SEE ALSO

=head2 Freshmeat API FAQ

http://freshmeat.net/faq/view/49/

=head2 Freshmeat XML-RPC API announcement

http://freshmeat.net/articles/view/1048/


=head1 DISCLAIMER

This is alpha software.   It does not contain necessary tests & checks yet ;-( 
    
However FreshMeat API is very simple & WWW::FreshMeat::API does work for me. 

I hope to make it beta status very shortly!


=head1 COPYRIGHT & LICENSE

Copyright 2009 Barry Walsh (Draegtun Systems Ltd), all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.







lib/WWW/FreshMeat/API.pm
package WWW::FreshMeat::API;
use Moose;

our $VERSION = '0.02';

has 'mock' => ( isa => 'Bool', is => 'ro', default => sub { 0 } );

with 'WWW::FreshMeat::API::Session', 
     'WWW::FreshMeat::API::Agent::XML::RPC',
     'WWW::FreshMeat::API::Pub';

no Moose;

1;


__END__


=head1 NAME

WWW::FreshMeat::API - inspect & update your freshmeat.net projects

=head1 VERSION

Version 0.02


=head1 SYNOPSIS

** IMPORTANT ANNOUNCEMENT **

Circa April 2009 Freshmeat relaunched their site and completely dropped the XML-RPC API :(

In place is a new Freshmeat API 3.0 using REST JSON (or XML).

Its not clear how stable this API is.    I'll look more closely at this in a few months 
time and perform necessary surgery on this module.

So note this module will not work in its present state until I make investigate the new API 
& changes source code changes.

Very sorry about this.

BW 19-Jul-2009

** END OF ANNOUCEMENT **


    use WWW::FreshMeat::API;
    
    my $fm = WWW::FreshMeat::API->new;
    
    $fm->login( username => 'user', password => 'pass' );
    
    say "Your ID for this API session (SID) is ", $fm->sid;

    for my $proj ( @{ $fm->fetch_project_list } ) {
        say "Project      ", $proj->{ projectname_full }; 
        say "- short name ", $proj->{ projectname_short };
        say "- status"    ", $proj->{ project_status };
        say "- version    ", $proj->{ project_version };
    }


=head1 DESCRIPTION

FreshMeat (http://freshmeat.net) provides a very simple XML-RPC API which allows a user to inspect what projects
the user as uploaded and also provides an update & withdrawal mechanism of the users projects.

Requirements....

    1) Must have a FreshMeat login & password

    2) Must have already loaded project onto http://freshmeat.net

For now see ./examples/freshmeat-submit.pl.

More details to follow.


=head1 EXPORT

None

=head1 METHODS

For session methods see L<WWW::FreshMeat::API::Session>

For Freshmeats public API which are mapped to methods see L<WWW::FreshMeat::API::Pub> & L<WWW::FreshMeat::API::Pub::V1_03>

=head1 ATTRIBUTES

=head2 mock

Constructor only testing attribute.   Not sure this will survive alpha so ignore for now!


=head1 AUTHOR

Barry Walsh, C<< <draegtun at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-freshmeat-api at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FreshMeat-API>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FreshMeat::API


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FreshMeat-API>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FreshMeat-API>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FreshMeat-API>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FreshMeat-API/>

=back

=head1 MOTIVATIONS

There certainly was!   More on that later ;-)


=head1 ACKNOWLEDGEMENTS

=head2 Moose

Thanks to Stevan Little and cast of thousands for the brilliant L<Moose> module. 
Using Moose really made developing this module simple and lots of fun!

=head2 XML::RPC

Thanks to Niek Albers, http://www.daansystems.com/ for his excellent L<XML::RPC> module.


=head1 SEE ALSO

=head2 Source Code

GitHub at  http://github.com/draegtun/www-freshmeat-api/tree/master

=head2 freshmeat-submit

Python script written by ESR no less!  http://freshmeat.net/projects/freshmeat-submit/

freshmeat-submit was written in 2003 and last updated 2004.   
I didn't find freshmeat-submit to be useful to me when writing WWW::FreshMeat::API.
Fingers crossed WWW::FreshMeat::API won't look as crufty as freshmeat-submit does after 5 and bit years ;-)


=head1 DISCLAIMER

This is alpha software.   It does not contain necessary tests & checks yet ;-( 
    
However FreshMeat API is very simple & WWW::FreshMeat::API does work for me. 

I hope to make it beta status very shortly!


=head1 COPYRIGHT & LICENSE

Copyright 2009 Barry Walsh (Draegtun Systems Ltd), all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

